--- testcases/itodo20.ics	2009-07-07 12:47:20.000000000 +0800
+++ testcases/itodo20.ics.funambol.tem	2009-07-07 13:57:58.000000000 +0800
@@ -32,8 +32,8 @@
 DUE;TZID=/softwarestudio.org/Tzfile/America/New_York:20090502T130000
 DTSTART;TZID=/softwarestudio.org/Tzfile/America/New_York:20090501T100000
 CLASS:PUBLIC
-PERCENT-COMPLETE:99
-STATUS:COMPLETED
+PERCENT-COMPLETE:78
+STATUS:IN-PROGRESS
 CATEGORIES:Business,Waiting
 COMPLETED:20090401T030000Z
 URL:http://foo.com
